x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -410,6 +425,32 @@ private Map<String, Object> buildUserMetadata(JWTClaimsSet claimsSet) {
         return Map.copyOf(metadata);
     }
 
+    private String buildTokenPrincipal(JWTClaimsSet jwtClaimsSet) {
```

Comment:
What all is the token principal used for ? 

I find the difference between `claims.principal` and the principal here abit confusing.  For example, I can set the  `claims.principal` to something compelety disconnected from the token principal. It's abit complicated (and may not be needed depending on exactly what the token principal is used for...) but had you considered using using `claims.principal` instead of (or in addition to) subject ?

Reply:
The `claims.principal` determines the user's username upon successful authentication. For example, for setting like`claims.principal: sub` and the `sub` claim has value `foo`. The username will be `foo`. 

But before the authentication process can complete, we also need an identifier for the incoming token. This is the token principal. The usages are:
* [Smart ordering for realm chain](https://github.com/elastic/elasticsearch/blob/a0bdd6582f21048e27bee619721f4e0fecaef929/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RealmsAuthenticator.java#L336) so that the realm that has previously succeeded in authenticating a token gets to authenticate the same token first irrespective to its designated realm order.
* Audit logging when authentication fails (and username is not available)
* Server side logging before authentication completes

ES supports multiple different tokens, including userpass token, oauth2 bearer token, x509 cert token, JWT bearer token, etc. Some tokens, e.g. userpass token, have their principals identical to the username. But they do not have to be. Ideally, the token principal should be unqiue across realms (or at least across realms accepting the same type of token) to avoid thrashing the cache of smart realm ordering. This is what the JWT's token principal tries to achieve here. Using the combination of `iss`, `aud` and `sub`, there is a much better chance for it to be unique across multiple JWT realms compare to using `claims.principal`.

Reply:
Thanks for the explanation !